### PR TITLE
[Feature][TPU host offload] add pallas dma swap ops

### PR DESCRIPTION
**Pull request by juncgu-google**
# Description

 1. add TPU_HBM <-> Host_DRAM swap operations using pallas DMA kernels, and the unit tests
 2. enforce KV cache layout to `NHD` in tpu_connector_local; currently, tpu jax runner uses `NHD` kv cache layout by default.
 3. add a wrap-function to choose between `jax.device_put()` and  pallas swap ops in tpu_connector_local , controlled by `TPU_OFFLOADING_SWAP_OP_TYPE` (jax, or pallas)

# Tests

`pytest -sv tests/kernels/host_dma_test.py`
`pytest -sv tests/distributed/cpu_offloading_kv_roundtrip_test.py`

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/distributed/cache_util.py [hunk 1]
```diff
@@ -3,15 +3,28 @@
...
+
+from tpu_inference.kernels.dma.host_dma import d2h_dma, h2d_dma
 from tpu_inference.logger import init_logger
 
 # Corresponds to the initial hash value
 NONE_HASH = 0
 
 logger = init_logger(__name__)
 
+# CopyBlocksOp = Callable[[
```
Comment (saikat-royc) [inline tpu_inference/distributed/cache_util.py]: please remove any commented code if not required in this PR

File: tpu_inference/distributed/tpu_connector_local.py [hunk 2]
```diff
@@ -107,14 +107,20 @@
...
+from .cache_util import CPU_OFFLOADING_SWAP_OP_TYPE, TokenProcessor, swap_ops
 from .local_cpu_backend import LocalCPUBackend
 
 EngineId = str
 ReqId = str
 
 logger = init_logger(__name__)
 
+# kv cache layout needed by cpu offloading mechanism
+REQUIRED_KV_CACHE_LAYOUT = "NHD"
```
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py:119]: please put a comment on what NHD means - e.g num tokens, heads, head dimensions

File: tpu_inference/distributed/tpu_connector_local.py [hunk 3]
```diff
@@ -725,6 +752,13 @@ def __init__(self, vllm_config: VllmConfig, connector: "TPUConnector"):
 
         self.runner: Optional[TPUModelRunner] = None
         self.mesh: Optional[Mesh] = None
+        self.swap_op_type = os.getenv("TPU_KV_OFFLOADING_SWAP_OP_TYPE",
```
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py]: for uniformity, let's use the existing prefix used in other env variables TPU_OFFLOAD_*

File: tpu_inference/distributed/tpu_connector_local.py [hunk 4]
```diff
@@ -107,14 +107,20 @@
...
 EngineId = str
 ReqId = str
 
 logger = init_logger(__name__)
 
+# kv cache layout needed by cpu offloading mechanism
+REQUIRED_KV_CACHE_LAYOUT = "NHD"
+
+# default swap op type
+DEFAULT_HOST_HBM_SWAP_OP_TYPE = "pallas"
```
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py]: We can default to 'jax' until pallas is stable?

File: tpu_inference/distributed/tpu_connector_local.py [hunk 5]
```diff
@@ -831,6 +865,12 @@ def _save_blocks_to_cpu(self, req_id: ReqId, full_block_ids: list[int],
                 layer_cache.reshape(-1, *layer_cache.shape[2:])
                 for layer_cache in kv_caches_on_cpu
             ]
+            # flat_kv_caches_on_cpu = [
```
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py]: please remove unused code. or if eishape is the preferred reshaping functionality, we can remove L864-867.

File: tpu_inference/distributed/tpu_connector_local.py [hunk 6]
```diff
@@ -1089,39 +1142,53 @@ def start_load_kv(self, fwd_ctx: "ForwardContext") -> None:
...
+            else:
+                padded_kv_on_cpu = final_kv_on_cpu
 
             # 4. Reshape data back to block format for the update operation.
             block_shaped_kv_on_cpu = [
                 layer_data.reshape(num_blocks_to_load, self.block_size,
                                    *layer_data.shape[1:])
                 for layer_data in padded_kv_on_cpu
             ]
+            # jax_einshape("(nb)...->nb...", layer_cache, b=block_size)
```
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py]: similar to other comments, remove commented code or replace the existing ones?

File: tpu_inference/distributed/cache_util.py [hunk 7]
```diff
@@ -75,3 +88,52 @@ def process_tokens(
...
+
+def swap_ops(
+    src_kv_cache: jax.Array,
+    out_sharding: Optional[jax.sharding.NamedSharding],
+    direction: Literal["h2d", "d2h"],
+    op_type: CPU_OFFLOADING_SWAP_OP_TYPE,
+) -> jax.Array:
+    if op_type == "jax":
+        return jax_swap_kv_cache(src_kv_cache, out_sharding, direction)
+    else:
```
Comment (saikat-royc) [inline tpu_inference/distributed/cache_util.py]: nit: else is not needed

if op_type == 'jax':
    return ...

return ...

File: tpu_inference/kernels/dma/host_dma.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
+import jax
```
Comment (saikat-royc) [inline tpu_inference/kernels/dma/host_dma.py:3]: please add the necessary license string

File: tpu_inference/distributed/tpu_connector_local.py [hunk 9]
```diff
@@ -1089,15 +1137,22 @@ def start_load_kv(self, fwd_ctx: "ForwardContext") -> None:
...
-                    padding = jnp.zeros((pad_width, *layer_data.shape[1:]),
-                                        dtype=layer_data.dtype)
-                    padded_kv_on_cpu.append(
-                        jnp.concatenate([layer_data, padding], axis=0))
-                else:
-                    padded_kv_on_cpu.append(layer_data)
+            pad_width = padded_token_len - num_tokens_to_load
+            if pad_width > 0:
+                pad_value = jnp.array(0, dtype=self.dtype)
+                padding_config = [
```
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py:1145]: for readability: please put a comment explaining that the first dimension which represents the tokens is being padded towards the end by pad_width, and why we are padding to avoid recompile?

**Comment by saikat-royc** [comment]
LGTM with comments

File: tests/distributed/cpu_offloading_kv_roundtrip_test.py [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+        keys_generator = worker.token_processor.process_tokens(
+            target_token_ids)
+        retrieved_chunks = []
+        for _, _, key in keys_generator:
+            cached_value = worker.cpu_backend.get(key)
+            assert cached_value is not None, f"Key {key} not found in CPU cache!"
+            assert len(
+                cached_value
+            ) == num_layers, f"cache_value layer: {len(cached_value)} != {num_layers}"
+            retrieved_chunks.append(cached_value[0])  # Get first layer
```
Comment (saikat-royc) [inline tests/distributed/cpu_offloading_kv_roundtrip_test.py:214]: we can compare all the layers content here
